### PR TITLE
[NVFuser] always use fallback if fusion fails

### LINKED ISSUE #74559: Improve NVFuser fallback performance
NVFuser fallbacks are slow.

For example, on autogen-69 from https://github.com/pytorch/benchmark/pull/801, we have roughly (for V100):
* 0.588 ms for eager
* 0.436 ms for nnc static
* 0.742 ms for nnc dynamic
* 33.6 ms for nvfuser, which encounters an error and hits a fallback

Part of this could be from caching the interpreter state / code as suggested below (to avoid costly graph copying and modification).

However, benchmarking shows that most of the runtime is actually spent in the fused implementation (i.e. before it fails). To fix this we could either:
* fix the fused implementation so it would fail faster, or
* if the fallback gets run, always just use the fallback instead of the fused implementation.

**original request below, from a comment on a PR**

We should really cache and save the `InterpreterState` code here. This fallback actually does get run in real networks, and this whole processes happens on each invocation

_Originally posted by @eellison in https://github.com/pytorch/pytorch/pull/74361#discussion_r831556105_

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #75983

[NVFuser] always fallback if fusion fails

1) remember when fusions fail; and on subsequent runs, always take the fallback.
2) during the first fallback, cache the Code object.

On autogen-69 from the nvfuser microbenchmarks (https://github.com/pytorch/benchmark/pull/801) this improved performanance as follows:
* Original (always attempt fusion): 25ms
* Always take fallback after first failure: 0.79ms
* Always take fallback + cache Code object: 0.62ms
* Eager: 0.58ms

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/manager.cpp [hunk 1]
```diff
@@ -52,6 +52,25 @@ namespace cuda {
 
 namespace {
 
+// Mark string attribute in alias-copy nodes to enable its implementation
+// in the fallback path.
+void enableAliasCopyNodes(const std::shared_ptr<Graph>& graph, Block* block) {
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/manager.cpp:57]: > we don't need this any more. I think we can use `revertAliasCopyOps`. Similar refactor should be done infallback graph used by fusion guard. implementation of xxxx_copy ops should be removed.

nitpick, I'll do a clean up afterwards. Do you mind putting a TODO here and copy-paste

File: torch/csrc/jit/codegen/cuda/manager.cpp [hunk 2]
```diff
@@ -110,6 +129,30 @@ class CudaFusionManager {
     return graph_cache_[kernel_id]->runGraphWithInputs(inputs);
   }
 
+  bool hasFallbackCode(int32_t kernel_id) {
+    std::lock_guard<std::mutex> guard(mutex_);
+    return graph_cache_.count(kernel_id);
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/manager.cpp:134]: shouldn't this be `fallback_cache_` instead?

**Review by jjsjann123 — CHANGES_REQUESTED** [review]

File: torch/csrc/jit/codegen/cuda/manager.cpp (hunk 2 above)
Comment (davidberard98) [reply torch/csrc/jit/codegen/cuda/manager.cpp]: Whoops, thanks for catching that!

File: torch/csrc/jit/codegen/cuda/manager.cpp (hunk 1 above)
Comment (davidberard98) [reply torch/csrc/jit/codegen/cuda/manager.cpp:62]: Added the TODO

**Review by jjsjann123 — APPROVED** [review]
LGTM.

**Comment by datumbox** [comment]
@davidberard98 We have started receiving "RuntimeError: required keyword attribute 'cache_id' is undefined" error messages at TorchVision after the landing of this PR. We are not 100% but we might be affected by it. Could you please have a look at: https://github.com/pytorch/vision/issues/5881

**Comment by davidberard98** [comment]
@pytorchmergebot revert this

@datumbox thanks for reporting it, I'll take a look

**Comment by jjsjann123** [comment]
> @pytorchmergebot revert this
> 
> @datumbox thanks for reporting it, I'll take a look

We could have fallback prior to getting a attr::cache_id set
https://github.com/pytorch/pytorch/blob/12a0b53f078052963afa8d5ee4442e49eeca90a2/torch/csrc/jit/codegen/cuda/manager.cpp#L209-L219

Fallback seems to be using that same id.
https://github.com/pytorch/pytorch/blob/12a0b53f078052963afa8d5ee4442e49eeca90a2/torch/csrc/jit/codegen/cuda/manager.cpp#L299

I think we needed to setup the cache_id for fallback in the catch here: https://github.com/pytorch/pytorch/blob/12a0b53f078052963afa8d5ee4442e49eeca90a2/torch/csrc/jit/codegen/cuda/manager.cpp#L226-L231

**Comment by jjsjann123** [comment]
> @davidberard98 We have started receiving "RuntimeError: required keyword attribute 'cache_id' is undefined" error messages at TorchVision after the landing of this PR. We are not 100% but we might be affected by it. Could you please have a look at: [pytorch/vision#5881](https://github.com/pytorch/vision/issues/5881)

Meanwhile, there seems to be a real failure here. Need to take a look

**Pull request closed without merging**